CA2000 does not handle "is not null"

### Analyzer

**Diagnostic ID**: [CA2000](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2000): `Dispose objects before losing scope`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 5.0.201](https://dotnet.microsoft.com/download/dotnet/5.0)

### Describe the bug

**Update: See repro below: https://github.com/dotnet/roslyn-analyzers/issues/4981#issuecomment-816461346**

CA2000 seems not to recognize `is not null` pattern but it recognize `!= null` in the following snippet:

```csharp
WebSocketClient? client = null; // `WebSocketClient` implements `IAsyncDisposable`

try
{
    client = new(uri, nameof(MyApiService));
    Result connectRes = await client.ConnectAsync(cancellationToken).ConfigureAwait(false);
    if (connectRes)
    {
        result = Result<IWebSocketClient>.Ok(client);
        client = null;
    }
    else result = connectRes;
}
finally
{
    if (client is not null) // <--- CHANGE THIS TO `if (client != null)` and the CA2000 warning is gone.
    {
        await client.DisposeAsync().ConfigureAwait(false);
    }
}
```

This is certainly not a shortest code one can come up but it is easy to understand I think.

### Steps To Reproduce

Please see the example above.

### Expected behavior

`if (client is not null)` does not lead to CA2000 being emitted.

### Actual behavior

`is not null` seems to be not supported but `!= null` is.

I have modified a test:

```csharp
        [Fact, WorkItem(2201, "https://github.com/dotnet/roslyn-analyzers/issues/2201")]
        public async Task NestedTryFinallyInTryCatch_NoDiagnostic()
        {
            await VerifyCS.VerifyAnalyzerAsync(@"
using System.IO;

class Test
{
    void M1()
    {
        try
        {
            var ms = new MemoryStream();
            try
            {
            }
            finally
            {
                if (ms is not null) {
                    ms.Dispose();
                }
            }
        }
        catch
        {
        }
    }
}");
```

but it returns

```
  Message: 
    Microsoft.CodeAnalysis.Testing.Verifiers.EqualWithMessageException : Context: Diagnostics of test state
    Mismatch between number of diagnostics returned, expected "0" actual "1"
    
    Diagnostics:
    // /0/Test0.cs(16,27): error CS8370: Feature 'not pattern' is not available in C# 7.3. Please use language version 9.0 or greater.
    DiagnosticResult.CompilerError("CS8370").WithSpan(16, 27, 16, 35).WithArguments("not pattern", "9.0"),
    
    
    Assert.Equal() Failure
    Expected: 0
    Actual:   1
```

So I can't probably fix it myself.

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DisposeObjectsBeforeLosingScopeTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DisposeObjectsBeforeLosingScopeTests.cs
@@ -12616,5 +12616,41 @@ class Program
         }
     }");
         }
+
+        [WorkItem(4981, "https://github.com/dotnet/roslyn-analyzers/issues/4981")]
+        [Theory]
+        [InlineData("ms != null")]
+        [InlineData("ms is not null")]
+        [InlineData("!(ms is null)")]
+        public async Task TryFinallyIsNotNull_NoDiagnostic(string condition)
+        {
+            var code = @$"
+using System.IO;
+
+class Test
+{{
+    void M1()
+    {{
+        MemoryStream ms = null;
+        try
+        {{
+            ms = new MemoryStream();
+        }}
+        finally
+        {{
+            if ({condition}) {{
+                ms.Dispose();
+            }}
+        }}
+    }}
+}}
+";
+            await new VerifyCS.Test
+            {
+                TestCode = code,
+                FixedCode = code,
+                LanguageVersion = CSharpLanguageVersion.CSharp9,
+            }.RunAsync();
+        }
     }
 }

--- a/src/Utilities/Compiler/Analyzer.Utilities.projitems
+++ b/src/Utilities/Compiler/Analyzer.Utilities.projitems
@@ -65,6 +65,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)Index.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)IsExternalInit.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Lightup\IMethodSymbolExtensions.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Lightup\INegatedPatternOperationWrapper.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Lightup\IOperationWrapper.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Lightup\ITypeSymbolExtensions.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Lightup\IFunctionPointerInvocationOperationWrapper.cs" />

--- a/src/Utilities/Compiler/Lightup/INegatedPatternOperationWrapper.cs
+++ b/src/Utilities/Compiler/Lightup/INegatedPatternOperationWrapper.cs
@@ -1,0 +1,51 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+#if HAS_IOPERATION
+
+namespace Analyzer.Utilities.Lightup
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.Operations;
+
+    [SuppressMessage("Performance", "CA1815:Override equals and operator equals on value types", Justification = "Not a comparable instance.")]
+    internal readonly struct INegatedPatternOperationWrapper : IOperationWrapper
+    {
+        internal const string WrappedTypeName = "Microsoft.CodeAnalysis.Operations.INegatedPatternOperation";
+        private static readonly Type? WrappedType = OperationWrapperHelper.GetWrappedType(typeof(INegatedPatternOperationWrapper));
+
+        private static readonly Func<IOperation, IPatternOperation> PatternAccessor = LightupHelpers.CreateOperationPropertyAccessor<IOperation, IPatternOperation>(WrappedType, nameof(Pattern), fallbackResult: null!);
+
+        private INegatedPatternOperationWrapper(IOperation operation)
+        {
+            WrappedOperation = operation;
+        }
+
+        public IOperation WrappedOperation { get; }
+        public ITypeSymbol? Type => this.WrappedOperation.Type;
+        public IPatternOperation Pattern => PatternAccessor(WrappedOperation);
+
+        public static INegatedPatternOperationWrapper FromOperation(IOperation operation)
+        {
+            if (operation == null)
+            {
+                return default;
+            }
+
+            if (!IsInstance(operation))
+            {
+                throw new InvalidCastException($"Cannot cast '{operation.GetType().FullName}' to '{WrappedTypeName}'");
+            }
+
+            return new INegatedPatternOperationWrapper(operation);
+        }
+
+        public static bool IsInstance(IOperation operation)
+        {
+            return operation != null && LightupHelpers.CanWrapOperation(operation, WrappedType);
+        }
+    }
+}
+
+#endif

--- a/src/Utilities/Compiler/Lightup/OperationWrapperHelper.cs
+++ b/src/Utilities/Compiler/Lightup/OperationWrapperHelper.cs
@@ -15,7 +15,8 @@ namespace Analyzer.Utilities.Lightup
 
         private static readonly ImmutableDictionary<Type, Type?> WrappedTypes = ImmutableDictionary.Create<Type, Type?>()
             .Add(typeof(IUsingDeclarationOperationWrapper), s_codeAnalysisAssembly.GetType(IUsingDeclarationOperationWrapper.WrappedTypeName))
-            .Add(typeof(IFunctionPointerInvocationOperationWrapper), s_codeAnalysisAssembly.GetType(IFunctionPointerInvocationOperationWrapper.WrappedTypeName));
+            .Add(typeof(IFunctionPointerInvocationOperationWrapper), s_codeAnalysisAssembly.GetType(IFunctionPointerInvocationOperationWrapper.WrappedTypeName))
+            .Add(typeof(INegatedPatternOperationWrapper), s_codeAnalysisAssembly.GetType(INegatedPatternOperationWrapper.WrappedTypeName));
 
         /// <summary>
         /// Gets the type that is wrapped by the given wrapper.

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs
@@ -8,6 +8,7 @@ using System.Diagnostics.CodeAnalysis;
 using System.Linq;
 using Analyzer.Utilities;
 using Analyzer.Utilities.Extensions;
+using Analyzer.Utilities.Lightup;
 using Analyzer.Utilities.PooledObjects;
 using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.CopyAnalysis;
 using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow.PointsToAnalysis;
@@ -1260,29 +1261,41 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
                 }
 
                 // Handle is pattern operations with constant pattern.
-                if (operation is IIsPatternOperation patternOperation &&
-                    patternOperation.Pattern is IConstantPatternOperation constantPattern)
+                if (operation is IIsPatternOperation isPatternOperation)
                 {
-                    if (constantPattern.Value.ConstantValue.HasValue)
+                    IPatternOperation patternOperation = isPatternOperation.Pattern;
+                    bool direct = true;
+
+                    if (INegatedPatternOperationWrapper.IsInstance(patternOperation))
                     {
-                        if (constantPattern.Value.ConstantValue.Value == null)
-                        {
-                            switch (pointsToValue.NullState)
-                            {
-                                case NullAbstractValue.Null:
-                                    inference.AlwaysSucceed = true;
-                                    break;
-
-                                case NullAbstractValue.NotNull:
-                                    inference.AlwaysFail = true;
-                                    break;
-                            }
-                        }
-
-                        return true;
+                        INegatedPatternOperationWrapper negatedPattern = INegatedPatternOperationWrapper.FromOperation(patternOperation);
+                        patternOperation = negatedPattern.Pattern;
+                        direct = false;
                     }
 
-                    return false;
+                    if (patternOperation is IConstantPatternOperation constantPattern)
+                    {
+                        if (constantPattern.Value.ConstantValue.HasValue)
+                        {
+                            if (constantPattern.Value.ConstantValue.Value == null)
+                            {
+                                switch (pointsToValue.NullState)
+                                {
+                                    case NullAbstractValue.Null:
+                                        inference.AlwaysSucceed = direct;
+                                        break;
+
+                                    case NullAbstractValue.NotNull:
+                                        inference.AlwaysFail = direct;
+                                        break;
+                                }
+                            }
+
+                            return true;
+                        }
+
+                        return false;
+                    }
                 }
 
                 if (targetType == null)

--- a/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs
+++ b/src/Utilities/FlowAnalysis/FlowAnalysis/Framework/DataFlow/DataFlowOperationVisitor.cs
@@ -1273,28 +1273,20 @@ namespace Microsoft.CodeAnalysis.FlowAnalysis.DataFlow
                         direct = false;
                     }
 
-                    if (patternOperation is IConstantPatternOperation constantPattern)
+                    if (patternOperation is IConstantPatternOperation { Value.ConstantValue: { HasValue: true, Value: null } })
                     {
-                        if (constantPattern.Value.ConstantValue.HasValue)
+                        switch (pointsToValue.NullState)
                         {
-                            if (constantPattern.Value.ConstantValue.Value == null)
-                            {
-                                switch (pointsToValue.NullState)
-                                {
-                                    case NullAbstractValue.Null:
-                                        inference.AlwaysSucceed = direct;
-                                        break;
+                            case NullAbstractValue.Null:
+                                inference.AlwaysSucceed = direct;
+                                break;
 
-                                    case NullAbstractValue.NotNull:
-                                        inference.AlwaysFail = direct;
-                                        break;
-                                }
-                            }
-
-                            return true;
+                            case NullAbstractValue.NotNull:
+                                inference.AlwaysFail = direct;
+                                break;
                         }
 
-                        return false;
+                        return true;
                     }
                 }
 

